### PR TITLE
Aesthetic Changes wrt 'size_t' data types.

### LINKED ISSUE #8: Correct Pruning for k >= 2
https://github.com/grf-labs/policytree/blob/master/src/tree_search.cpp#L334

At this point, pruning should be incorrect since we treat the non pruned optimal nodes differently from the pruned, and in the case that the pruned node is more optimal than two non pruned nodes it splits into, we will report the non pruned always.

### PR DESCRIPTION
**Pull request by kanodiaayush**
1. Do not set or compare size_t with negative values
2. Leaf nodes are checked by testing whether their children are nullptr

Fixes #10

### CODE REVIEW

File: src/tree_search.cpp [hunk 1]
```diff
@@ -341,15 +341,15 @@ std::unique_ptr<Node> find_best_split(const std::vector<flat_set>& sorted_sets,
       return level_zero_learning(sorted_sets, data);
     } else {
           // "pruning", the recursive case (same action in both leaves):
-      if ((best_left_child->action_id != -1) &&
+      if ((best_left_child->left_child == nullptr && best_right_child->left_child == nullptr) &&
               (best_left_child->action_id == best_right_child->action_id)) {
         double leaf_reward = best_left_child->reward + best_right_child->reward;
         size_t leaf_action = best_left_child->action_id;
-        best_ans_as_leaf = std::unique_ptr<Node> (new Node(-1, -1, leaf_reward, leaf_action));
+        best_ans_as_leaf = std::unique_ptr<Node> (new Node(0, -1, leaf_reward, leaf_action));
```
Comment (erikcs) [inline src/tree_search.cpp]: There are still other places to make this change: like :195

Comment (erikcs) [reply src/tree_search.cpp]: And please add a more descriptive PR title and description, e.g. that this is an "aesthetic" fix

File: src/tree_search.cpp [hunk 2]
```diff
@@ -341,15 +341,15 @@ std::unique_ptr<Node> find_best_split(const std::vector<flat_set>& sorted_sets,
       return level_zero_learning(sorted_sets, data);
     } else {
           // "pruning", the recursive case (same action in both leaves):
-      if ((best_left_child->action_id != -1) &&
+      if ((best_left_child->left_child == nullptr && best_right_child->left_child == nullptr) &&
```
Comment (erikcs) [inline src/tree_search.cpp]: And can you make a note of this in the Node struct in tree_search.h? That the convention is to set them to zero...

And what about making left_child == nullptr a method of Node? (bool is_leaf)

Comment (kanodiaayush) [reply src/tree_search.cpp]: I did consider is_leaf and I think that is a great idea. I'll do that.

File: src/tree_search.cpp (hunk 1 above)
Comment (kanodiaayush) [reply src/tree_search.cpp]: Upcoming..

**Review by erikcs — APPROVED** [review]
Thanks, looks good! I thought functions defined in the class template were inlined automatically. And the PR post is still outdated: it is in quotes and referring to 8 and 9...?

**Comment by kanodiaayush** [comment]
> Thanks, looks good! I thought functions defined in the class template were inlined automatically. And the PR post is still outdated: it is in quotes and referring to 8 and 9...?

Fixed! I'm not sure the inline is automatic - maybe quote a reference? I'm looking here -- [here](https://stackoverflow.com/questions/10535667/does-it-make-any-sense-to-use-inline-keyword-with-templates)

**Comment by erikcs** [comment]
https://isocpp.org/wiki/faq/inline-functions#inline-member-fns-more

**Comment by kanodiaayush** [comment]
Thanks, I was looking at something incorrect. Merged.

**Pull request merged**